### PR TITLE
Add SARS tdlearning back to lib

### PR DESCRIPTION
**Pull request by jeremiahpslewis**
- One of a couple pull requests on the way to getting plain vanilla q-learning working
- Drops optimiser from `TabularApproximator`, too many unforced errors with it specified
- `Approximator` becomes `FluxModelApproximator` for clarity
- `TabularApproximator` and `FluxModelApproximator` are independent subtypes of `AbstractLearner`

PR Checklist

- [ ] Update NEWS.md?
- [ ] Unit tests for all structs / functions?
- [ ] Integration and correctness tests using a simple env?
- [ ] PR Review?
- [ ] Add or update documentation?
- [ ] Write docstrings for new methods?

### CODE REVIEW

File: src/ReinforcementLearningCore/test/policies/q_based_policy.jl [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        explorer = EpsilonGreedyExplorer(0.1)
+        policy = QBasedPolicy(q_approx, explorer)
+        s = PreActStage()
+        trajectory = Trajectory(
+            CircularArraySARTSTraces(; capacity = 1),
+            BatchSampler(1),
+            InsertSampleRatioController(n_inserted = -1),
+        )
+        RLBase.optimise!(policy, s, trajectory)
+        # Add assertions here
```
Comment (HenriDeh) [inline src/ReinforcementLearningCore/test/policies/q_based_policy.jl:53]: Same here

File: src/ReinforcementLearningFarm/src/algorithms/tabular/td_learner.jl [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+Use temporal-difference method to estimate state value or state-action value.
+
+# Fields
+- `approximator` is `<:TabularApproximator`.
+- `γ=1.0`, discount rate.
+- `method`: only `:SARS` (Q-learning) is supported for the time being.
+- `n=0`: the number of time steps used minus 1.
+"""
+Base.@kwdef mutable struct TDLearner{M,A} <: AbstractLearner where {A<:TabularApproximator,M<:Symbol}
```
Comment (HenriDeh) [inline src/ReinforcementLearningFarm/src/algorithms/tabular/td_learner.jl:21]: Should we alias TDLearner{:SARS} to SARSLearner ?

File: src/ReinforcementLearningFarm/src/algorithms/tabular/td_learner.jl [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+end
+
+RLCore.forward(L::TDLearner, s) = RLCore.forward(L.approximator, s)
+RLCore.forward(L::TDLearner, s, a) = RLCore.forward(L.approximator, s, a)
+
+Q(app::TabularApproximator, s, a) = RLCore.forward(app, s, a)
+Q(app::TabularApproximator, s) = RLCore.forward(app, s)
+
+"""
+    Q!(app::TabularApproximator, s::Int, s_plus_one::Int, a::Int, α::Float64, π_::Float64, γ::Float64)
```
Comment (HenriDeh) [inline src/ReinforcementLearningFarm/src/algorithms/tabular/td_learner.jl:42]: I'm not a fan of naming the Bellman update function `Q!`. I'd rather go with a name which is more descriptive of the function. Like `bellman_update!` or `update_Q!`, or whatever.

File: src/ReinforcementLearningFarm/src/algorithms/tabular/td_learner.jl [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    a::I3,
+    π_::F1,
+    γ::Float64, # discount factor
+) where {I1<:Integer,I2<:Integer,I3<:Integer,F1<:AbstractFloat}
+    # Q-learning formula following https://github.com/JuliaPOMDP/TabularTDLearning.jl/blob/25c4d3888e178c51ed1ff448f36b0fcaf7c1d8e8/src/q_learn.jl#LL63C26-L63C95
+    # Terminology following https://en.wikipedia.org/wiki/Q-learning
+    estimate_optimal_future_value = maximum(Q(approx, s_plus_one))
+    current_value = Q(approx, s, a)
+    raw_q_value = (π_ + γ * estimate_optimal_future_value - current_value) # Discount factor γ is applied here
+    q_value_updated = Flux.Optimise.apply!(approx.optimiser_state, :learning, [raw_q_value])[] # adust according to optimiser learning rate
```
Comment (HenriDeh) [inline src/ReinforcementLearningFarm/src/algorithms/tabular/td_learner.jl:59]: I think this is the old Flux api

File: src/ReinforcementLearningFarm/src/algorithms/tabular/td_learner.jl [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    s_plus_one::I2,
+    a::I3,
+    π_::F1,
+    γ::Float64, # discount factor
+) where {I1<:Integer,I2<:Integer,I3<:Integer,F1<:AbstractFloat}
+    # Q-learning formula following https://github.com/JuliaPOMDP/TabularTDLearning.jl/blob/25c4d3888e178c51ed1ff448f36b0fcaf7c1d8e8/src/q_learn.jl#LL63C26-L63C95
+    # Terminology following https://en.wikipedia.org/wiki/Q-learning
+    estimate_optimal_future_value = maximum(Q(approx, s_plus_one))
+    current_value = Q(approx, s, a)
+    raw_q_value = (π_ + γ * estimate_optimal_future_value - current_value) # Discount factor γ is applied here
```
Comment (HenriDeh) [inline src/ReinforcementLearningFarm/src/algorithms/tabular/td_learner.jl:58]: is pi_ the reward? If so, let's name it r

File: src/ReinforcementLearningFarm/src/algorithms/tabular/td_learner.jl (hunk 2 above)
Comment (jeremiahpslewis) [reply src/ReinforcementLearningFarm/src/algorithms/tabular/td_learner.jl:21]: Would this follow standard naming conventions? Or did we use to have a SARSLearner? My only concern is that the TD reference is then dropped / missing from the name

Comment (HenriDeh) [reply src/ReinforcementLearningFarm/src/algorithms/tabular/td_learner.jl:21]: SARS is a TD method so it kind of goes without saying. But we don't have to do this right now. This is easy to do in a later PR.

**Review by HenriDeh — APPROVED** [review]
LGTM. Good idea naming the FluxApproximator. One day this could be moving to a Flux extension.

**Pull request merged**